lib/Netspoc/Compiler/Pass1.pm

```diff
@@ -7867,110 +8023,167 @@ sub distribute_nat1 {
+# Purpose:    Performs a depth first search to distribute specified NAT tag
+#             to reachable domains where NAT tag is active; checks whether
+#             NAT declarations are applied correctly. 
+# Parameters: $domain: Domain the depth first search proceeds from.
+#             $nat_tag: NAT tag that is to be distributed.
+#             $nat_tag2multinat_def: Lookup hash for elements with more than
+#                 one NAT tag specified.
+#             $in_router: Router $domain was entered from.
+# Results:    All domains, where NAT tag is active contain $nat_tag in their
+#             $nat_set variable.
+# Returns:    undef on success, array reference of routers, if invalid
+#             path was found in loop.
+sub distribute_nat1 {
+    my ($domain, $nat_tag, $nat_tag2multinat_def, $in_router) = @_;
+#    debug "nat:$nat_tag at $domain->{name} from $in_router->{name}";
+
+    # Loop found or domain was processed by earlier call of distribute_nat.
+    my $nat_set = $domain->{nat_set};    
+    return if $nat_set->{$nat_tag};
+
+    # Perform checks before $nat_tag is added.
+    check_for_multinat_errors($nat_tag2multinat_def, $nat_set,
+                              $nat_tag, $domain);
+    check_for_translation_errors($domain, $nat_tag, $in_router);
+    $nat_set->{$nat_tag} = 1;
 
     # Activate loop detection.
     local $in_router->{active_path} = 1;
-    local $domain->{active_path}    = 1;
 
-    # Distribute NAT tag to adjacent NAT domains.
+    # Find adjacent domains with active $nat_tag to proceed traversal.
     for my $router (@{ $domain->{routers} }) {
         next if $router eq $in_router;
-        my $in_nat_tags = $router->{nat_tags}->{$domain};
 
-        # Found another interface with same NAT binding.
-        # This stops effect of current NAT tag.
+        # $nat_tag is deactivated at routers domain facing interface.
+        my $in_nat_tags = $router->{nat_tags}->{$domain};
         next if grep { $_ eq $nat_tag } @$in_nat_tags;
 
-        # Traverse loop twice to prevent inherited errors.
-        # Check for recursive and duplicate NAT.
-        for my $out_domain (@{ $router->{nat_domains} }) {
-            next if $out_domain eq $domain;
-            my $out_nat_tags = $router->{nat_tags}->{$out_domain};
-
-            # Must not apply one NAT tag multiple times in a row.
-            if (grep { $_ eq $nat_tag } @$out_nat_tags) {
-
-                # Check for recursive NAT in loop.
-                if ($router->{active_path}) {
-
-                    # Abort traversal and start collecting path.
-                    return [$router];
-                }
-                err_msg(
-                    "nat:$nat_tag is applied twice between",
-                    " $in_router->{name} and $router->{name}"
-                );
-            }
-        }
+        my $loop_error = check_for_proper_NAT_binding($domain, $in_router,
```

Comment:
Überlege, die Logik umzudrehen: 1=ok, 0=Fehler,
dann kann man schreiben check_for_proper..(..) or return [$router]
Außerdem finde ich meinen originalen Kommentar besser:
"# Abort traversal and start collecting path.".